File: moves.cpp
#include "moves.h"
#include <algorithm>
#include "utils.h"
string R(char green[3][3], char white[3][3], char yellow[3][3], char red[3][3], char orange[3][3], char blue[3][3]) {

    R_prime(green, white, yellow, red, orange, blue);
    R_prime(green, white, yellow, red, orange, blue);
    R_prime(green, white, yellow, red, orange, blue);

    return "R";
}

string L(char green[3][3], char white[3][3], char yellow[3][3], char red[3][3], char orange[3][3], char blue[3][3]) {

    char temp[3];

    for (int i = 0; i < 3; i++)
        temp[i] = yellow[i][0];

    for (int i = 0; i < 3; i++)
        yellow[i][0] = blue[2 - i][2];

    for (int i = 0; i < 3; i++)
        blue[i][2] = white[2 - i][0];

    for (int i = 0; i < 3; i++)
        white[i][0] = green[i][0];

    for (int i = 0; i < 3; i++)
        green[i][0] = temp[i];

    // rotate red face clockwise
    for (int i = 0; i < 3; i++)
        for (int j = i; j < 3; j++)
            swap(red[i][j], red[j][i]);

    for (int i = 0; i < 3; i++)
        reverse(red[i], red[i] + 3);

    return "L";
}

//gggggggggwwwwwwwwwyyyyyyyyyrrrrrrrrrooooooooobbbbbbbbb
string R_prime(char green[3][3], char white[3][3], char yellow[3][3], char red[3][3], char orange[3][3], char blue[3][3]) {
    char temp[3];

    for (int i = 0; i < 3; i++)
        temp[i] = yellow[i][2];

    for (int i = 0; i < 3; i++)
        yellow[i][2] = blue[2 - i][0];

    // blue <- white (reversed)
    for (int i = 0; i < 3; i++)
        blue[i][0] = white[2 - i][2];

    // white <- green
    for (int i = 0; i < 3; i++)
        white[i][2] = green[i][2];

    // green <- old yellow
    for (int i = 0; i < 3; i++)
        green[i][2] = temp[i];

    // rotate orange face counterclockwise
    for (int i = 0; i < 3; i++)
        for (int j = i; j < 3; j++)
            swap(orange[i][j], orange[j][i]);

    // reverse columns (not rows)
    for (int j = 0; j < 3; j++)
        for (int i = 0; i < 3 / 2; i++)
            swap(orange[i][j], orange[2 - i][j]);


    return "R'";
}

string L_prime(char green[3][3], char white[3][3], char yellow[3][3], char red[3][3], char orange[3][3], char blue[3][3]) {
    L(green, white, yellow, red, orange, blue);
    L(green, white, yellow, red, orange, blue);
    L(green, white, yellow, red, orange, blue);
    return "L'";
}
string F(char green[3][3], char white[3][3], char yellow[3][3], char red[3][3], char orange[3][3], char blue[3][3]) {

    // rotate green face clockwise
    char temp_face[3][3];
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            temp_face[i][j] = green[2 - j][i];

    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            green[i][j] = temp_face[i][j];

    char temp[3];

    for (int i = 0; i < 3; i++)
        temp[i] = yellow[2][i];

    for (int i = 0; i < 3; i++)
        yellow[2][i] = red[2 - i][2];

    for (int i = 0; i < 3; i++)
        red[i][2] = white[0][i];

    for (int i = 0; i < 3; i++)
        white[0][i] = orange[2 - i][0];

    for (int i = 0; i < 3; i++)
        orange[i][0] = temp[i];

    return "F";
}

string F_prime(char green[3][3], char white[3][3], char yellow[3][3], char red[3][3], char orange[3][3], char blue[3][3]) {
    F(green, white, yellow, red, orange, blue);
    F(green, white, yellow, red, orange, blue);
    F(green, white, yellow, red, orange, blue);
    return "F'";
}

string U(char green[3][3], char white[3][3], char yellow[3][3], char red[3][3], char orange[3][3], char blue[3][3]) {
    // Rotate yellow face clockwise
    char temp_face[3][3];
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            temp_face[i][j] = yellow[2 - j][i];
        }
    }
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            yellow[i][j] = temp_face[i][j];
        }
    }

    // Adjust adjacent faces
    char temp_row[3];
    for (int i = 0; i < 3; i++) {
        temp_row[i] = red[0][i];
    }
    for (int i = 0; i < 3; i++) {
        red[0][i] = green[0][i];
    }
    for (int i = 0; i < 3; i++) {
        green[0][i] = orange[0][i];
    }
    for (int i = 0; i < 3; i++) {
        orange[0][i] = blue[0][i];
    }
    for (int i = 0; i < 3; i++) {
        blue[0][i] = temp_row[i];
    }
    return "U";
}

string U_prime(char green[3][3], char white[3][3], char yellow[3][3], char red[3][3], char orange[3][3], char blue[3][3]) {
    U(green, white, yellow, red, orange, blue);
    U(green, white, yellow, red, orange, blue);
    U(green, white, yellow, red, orange, blue);
    return "U'";
}

string x(char green[3][3], char white[3][3], char yellow[3][3], char red[3][3], char orange[3][3], char blue[3][3]) {

    char temp[3][3];

    // save yellow
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            temp[i][j] = yellow[i][j];

    // up <- front
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            yellow[i][j] = green[i][j];

    // front <- down
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            green[i][j] = white[i][j];

    // down <- back (180°)
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            white[i][j] = blue[2 - i][2 - j];

    // back <- old up (180°)
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            blue[i][j] = temp[2 - i][2 - j];

    // rotate right clockwise
    for (int i = 0; i < 3; i++)
        for (int j = i; j < 3; j++)
            swap(orange[i][j], orange[j][i]);

    for (int i = 0; i < 3; i++)
        reverse(orange[i], orange[i] + 3);

    // rotate left counterclockwise
    for (int i = 0; i < 3; i++)
        for (int j = i; j < 3; j++)
            swap(red[i][j], red[j][i]);

    for (int j = 0; j < 3; j++)
        for (int i = 0; i < 3 / 2; i++)
            swap(red[i][j], red[2 - i][j]);

    return "x";
}

string x_prime(char green[3][3], char white[3][3], char yellow[3][3], char red[3][3], char orange[3][3], char blue[3][3]) {
    // Rotate entire cube around x-axis counterclockwise (equivalent to three x moves)
    x(green, white, yellow, red, orange, blue);
    x(green, white, yellow, red, orange, blue);
    x(green, white, yellow, red, orange, blue);
    return "x'";
}

string y_prime(char green[3][3], char white[3][3], char yellow[3][3], char red[3][3], char orange[3][3], char blue[3][3]) {

    char temp[3][3];

    // save green
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            temp[i][j] = green[i][j];

    // front <- left
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            green[i][j] = red[i][j];

    // left <- back
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            red[i][j] = blue[i][j];

    // back <- right
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            blue[i][j] = orange[i][j];

    // right <- old front
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            orange[i][j] = temp[i][j];

    // rotate up (yellow) counterclockwise
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            temp[i][j] = yellow[j][2 - i];

    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            yellow[i][j] = temp[i][j];

    // rotate down (white) clockwise
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            temp[i][j] = white[2 - j][i];

    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            white[i][j] = temp[i][j];


    return "y'";
}

string y(char green[3][3], char white[3][3], char yellow[3][3], char red[3][3], char orange[3][3], char blue[3][3]) {
    // Rotate entire cube around y-axis counterclockwise (equivalent to three y moves)
    y_prime(green, white, yellow, red, orange, blue);
    y_prime(green, white, yellow, red, orange, blue);
    y_prime(green, white, yellow, red, orange, blue);
    return "y";
}

string M_prime(char green[3][3], char white[3][3], char yellow[3][3], char red[3][3], char orange[3][3], char blue[3][3])
{
    char temp[3];

    // save yellow middle column
    for (int i = 0; i < 3; i++)
        temp[i] = yellow[i][1];

    // yellow <- green
    for (int i = 0; i < 3; i++)
        yellow[i][1] = green[i][1];

    // green <- white
    for (int i = 0; i < 3; i++)
        green[i][1] = white[i][1];

    // white <- blue (reversed)
    for (int i = 0; i < 3; i++)
        white[i][1] = blue[2 - i][1];

    // blue <- old yellow (reversed)
    for (int i = 0; i < 3; i++)
        blue[i][1] = temp[2 - i];

    return "M'";
}

string M(char green[3][3], char white[3][3], char yellow[3][3], char red[3][3], char orange[3][3], char blue[3][3]) {
    M_prime(green, white, yellow, red, orange, blue);
    M_prime(green, white, yellow, red, orange, blue);
    M_prime(green, white, yellow, red, orange, blue);

    return "M";
}

File: moves.js
// Moves for Rubik's Cube

function R(front, down, up, left, right, back) {
    R_prime(front, down, up, left, right, back);
    R_prime(front, down, up, left, right, back);
    R_prime(front, down, up, left, right, back);
    return "R";
}

function L(front, down, up, left, right, back) {
    const temp = [top[0][0], top[1][0], top[2][0]];

    for (let i = 0; i < 3; i++)
        top[i][0] = back[2 - i][2];

    for (let i = 0; i < 3; i++)
        back[i][2] = down[2 - i][0];

    for (let i = 0; i < 3; i++)
        down[i][0] = front[i][0];

    for (let i = 0; i < 3; i++)
        front[i][0] = temp[i];

    // rotate left face clockwise
    for (let i = 0; i < 3; i++) {
        for (let j = i; j < 3; j++) {
            [left[i][j], left[j][i]] = [left[j][i], left[i][j]];
        }
    }

    for (let i = 0; i < 3; i++) {
        red[i].reverse();
    }

    return "L";
}

function R_prime(front, down, up, left, right, back) {
    const temp = [up[0][2], up[1][2], up[2][2]];

    for (let i = 0; i < 3; i++)
        up[i][2] = back[2 - i][0];

    // blue <- white (reversed)
    for (let i = 0; i < 3; i++)
        back[i][0] = down[2 - i][2];

    // white <- green
    for (let i = 0; i < 3; i++)
        down[i][2] = front[i][2];

    // green <- old yellow
    for (let i = 0; i < 3; i++)
        front[i][2] = temp[i];

    // rotate orange face counterclockwise
    for (let i = 0; i < 3; i++) {
        for (let j = i; j < 3; j++) {
            [right[i][j], right[j][i]] = [right[j][i], right[i][j]];
        }
    }

    // reverse columns (not rows)
    for (let j = 0; j < 3; j++) {
        for (let i = 0; i < Math.floor(3 / 2); i++) {
            [right[i][j], right[2 - i][j]] = [right[2 - i][j], right[i][j]];
        }
    }

    return "R'";
}

function L_prime(front, down, up, left, right, back) {
    L(front, down, up, left, right, back);
    L(front, down, up, left, right, back);
    L(front, down, up, left, right, back);
    return "L'";
}

function F(front, down, up, left, right, back) {
    // rotate green face clockwise
    const temp_face = Array(3).fill(0).map(() => Array(3));
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            temp_face[i][j] = front[2 - j][i];
        }
    }

    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            front[i][j] = temp_face[i][j];
        }
    }

    const temp = [down[2][0], down[2][1], down[2][2]];

    for (let i = 0; i < 3; i++)
        front[2][i] = right[2 - i][2];

    for (let i = 0; i < 3; i++)
        right[i][2] = up[0][i];

    for (let i = 0; i < 3; i++)
        up[0][i] = left[2 - i][0];

    for (let i = 0; i < 3; i++)
        left[i][0] = temp[i];

    return "F";
}

function F_prime(front, down, up, left, right, back) {
    F(front, down, up, left, right, back);
    F(front, down, up, left, right, back);
    F(front, down, up, left, right, back);
    return "F'";
}

function U(front, down, up, left, right, back) {
    // Rotate yellow face clockwise
    const temp_face = Array(3).fill(0).map(() => Array(3));
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            temp_face[i][j] = up[2 - j][i];
        }
    }
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            up[i][j] = temp_face[i][j];
        }
    }

    // Adjust adjacent faces
    const temp_row = [...left[0]];
    for (let i = 0; i < 3; i++) {
        right[0][i] = front[0][i];
    }
    for (let i = 0; i < 3; i++) {
        front[0][i] = back[0][i];
    }
    for (let i = 0; i < 3; i++) {
        back[0][i] = temp_row[i];
    }
    for (let i = 0; i < 3; i++) {
        blue[0][i] = temp_row[i];
    }
    return "U";
}

function U_prime(front, down, up, left, right, back) {
    U(front, down, up, left, right, back);
    U(front, down, up, left, right, back);
    U(front, down, up, left, right, back);
    return "U'";
}

function x(front, down, up, left, right, back) {
    const temp = Array(3).fill(0).map(() => Array(3));

    // save yellow
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            temp[i][j] = up[i][j];
        }
    }

    // up <- front
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            up[i][j] = front[i][j];
        }
    }

    // front <- down
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            front[i][j] = down[i][j];
        }
    }

    // down <- back (180°)
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            down[i][j] = back[2 - i][2 - j];
        }
    }

    // back <- old up (180°)
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            back[i][j] = temp[2 - i][2 - j];
        }
    }

    // rotate right clockwise
    for (let i = 0; i < 3; i++) {
        for (let j = i; j < 3; j++) {
            [right[i][j], right[j][i]] = [right[j][i], right[i][j]];
        }
    }

    for (let i = 0; i < 3; i++) {
        left[i].reverse();
    }

    // rotate left counterclockwise
    for (let i = 0; i < 3; i++) {
        for (let j = i; j < 3; j++) {
            [left[i][j], left[j][i]] = [left[j][i], left[i][j]];
        }
    }

    for (let j = 0; j < 3; j++) {
        for (let i = 0; i < Math.floor(3 / 2); i++) {
            [right[i][j], right[2 - i][j]] = [right[2 - i][j], right[i][j]];
        }
    }

    return "x";
}

function x_prime(front, down, up, left, right, back) {
    x(front, down, up, left, right, back);
    x(front, down, up, left, right, back);
    x(front, down, up, left, right, back);
    return "x'";
}

function y_prime(front, down, up, left, right, back) {
    const temp = Array(3).fill(0).map(() => Array(3));

    // save green
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            temp[i][j] = front[i][j];
        }
    }

    // front <- left
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            front[i][j] = left[i][j];
        }
    }

    // left <- back
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            left[i][j] = back[i][j];
        }
    }

    // back <- right
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            back[i][j] = right[i][j];
        }
    }

    // right <- old front
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            right[i][j] = temp[i][j];
        }
    }

    // rotate up (up) counterclockwise
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            temp[i][j] = up[j][2 - i];
        }
    }

    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            top[i][j] = temp[i][j];
        }
    }

    // rotate down (down) clockwise
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            temp[i][j] = down[2 - j][i];
        }
    }

    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            down[i][j] = temp[i][j];
        }
    }

    return "y'";
}

function y(front, down, up, left, right, back) {
    y_prime(front, down, up, left, right, back);
    y_prime(front, down, up, left, right, back);
    y_prime(front, down, up, left, right, back);
    return "y";
}

function M_prime(front, down, up, left, right, back) {
    const temp = [up[0][1], up[1][1], up[2][1]];

    // up <- front
    for (let i = 0; i < 3; i++)
        up[i][1] = front[i][1];

    // front <- down
    for (let i = 0; i < 3; i++)
        front[i][1] = down[i][1];

    // down <- back (reversed)
    for (let i = 0; i < 3; i++)
        down[i][1] = back[2 - i][1];

    // back <- old up
    for (let i = 0; i < 3; i++)
        back[i][1] = temp[i];

    return "M'";
}

function M(front, down, up, left, right, back) {
    M_prime(front, down, up, left, right, back);
    M_prime(front, down, up, left, right, back);
    M_prime(front, down, up, left, right, back);
    return "M";
}
